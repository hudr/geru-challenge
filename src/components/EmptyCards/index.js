import React, { Component, Fragment } from 'react'

import { StyledEmptyContainer, NotesImg, Subtitle } from './styled'

import emptyImage from '../../assets/images/nodata.svg'

class EmptyCards extends Component {
  render() {
    return (
      <Fragment>
        <StyledEmptyContainer>
          <NotesImg src={emptyImage} alt='emptyimage' />
          <Subtitle>Your todo list is empty :)</Subtitle>
        </StyledEmptyContainer>
      </Fragment>
    )
  }
}

export default EmptyCards
