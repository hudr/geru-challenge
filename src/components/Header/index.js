import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import * as TodoActions from '../../redux/actions/todo'

import { StyledContainer, Title, CreateButton } from './styled'

class Header extends Component {
  render() {
    const { openModal } = this.props
    return (
      <StyledContainer>
        <Title>Geru Challenge</Title>
        <CreateButton onClick={() => openModal('post')}>Create</CreateButton>
      </StyledContainer>
    )
  }
}

const mapDispatchToProps = dispatch => bindActionCreators(TodoActions, dispatch)

export default connect(
  null,
  mapDispatchToProps
)(Header)

Header.propTypes = {
  openModal: PropTypes.func
}

export { Header }
