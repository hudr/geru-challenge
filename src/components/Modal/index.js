import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'

import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import * as TodoActions from '../../redux/actions/todo'

import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'
import { withStyles } from '@material-ui/core/styles'
import withMobileDialog from '@material-ui/core/withMobileDialog'
import TextField from '@material-ui/core/TextField'
import Radio from '@material-ui/core/Radio'
import RadioGroup from '@material-ui/core/RadioGroup'
import FormControlLabel from '@material-ui/core/FormControlLabel'

import { MuiPickersUtilsProvider, DateTimePicker } from '@material-ui/pickers'
import DateFnsUtils from '@date-io/date-fns'

import TagsInput from 'react-tagsinput'
import 'react-tagsinput/react-tagsinput.css'

import { TagsContainer } from './styled'

const useStyles = theme => ({
  textField: {
    marginLeft: theme.spacing(0),
    marginRight: theme.spacing(0)
  },
  dateField: {
    marginTop: theme.spacing(4)
  },
  radioButton: {
    marginTop: theme.spacing(2)
  }
})

class Modal extends Component {
  constructor(props) {
    super(props)
    this.state = {
      id: props.currentTodo.id || '',
      title: props.currentTodo.title || '',
      description: props.currentTodo.description || '',
      dueDate: props.currentTodo.dueDate || new Date(),
      createDate: props.currentTodo.createDate || new Date(),
      estTime: props.currentTodo.estTime || '',
      done: props.currentTodo.done || 'false',
      tags: props.currentTodo.tags || []
    }

    this.handleTagsChange = this.handleTagsChange.bind(this)
  }

  handleDateChange = date => {
    this.setState({
      dueDate: date
    })
  }

  handleTagsChange(tags) {
    this.setState({ tags })
  }

  handleStatusChange(event) {
    this.setState({ done: event.target.value })
  }

  handleInputOnChange = e => {
    this.setState({ [e.target.name]: e.target.value })
  }

  handleSubmit = async event => {
    event.preventDefault()

    const { modalType, addTodo, updateTodo } = this.props

    const {
      id,
      title,
      description,
      createDate,
      dueDate,
      estTime,
      done,
      tags
    } = this.state

    const currentTodo = {
      id,
      title,
      description,
      createDate,
      dueDate,
      estTime,
      done,
      tags
    }

    if (modalType === 'put') {
      await updateTodo(currentTodo)
    } else {
      await addTodo(currentTodo)
    }

    return this.setState({
      title: '',
      description: '',
      estTime: '',
      dueDate,
      tags: []
    })
  }

  render() {
    const {
      fullScreen,
      classes,
      handleClose,
      isModalOpen,
      closeModal,
      modalType
    } = this.props

    const { title, description, estTime, dueDate, done } = this.state
    return (
      <Fragment>
        <Dialog
          fullScreen={fullScreen}
          open={isModalOpen}
          onClose={handleClose}
          aria-labelledby='responsive-dialog-title'
        >
          <DialogTitle id='responsive-dialog-title'>
            {'Fill your todo info :)'}
          </DialogTitle>
          <DialogContent>
            <DialogContentText>
              Don't forget to add tags, they're important to remember what you
              need to do. Tag names can't be the same.
            </DialogContentText>
            <form
              onSubmit={this.handleSubmit.bind(this)}
              onChange={this.handleInputOnChange}
              id='todoinfo'
              autoComplete='off'
            >
              <TextField
                label='Title'
                required
                name='title'
                placeholder='e.g.: Homework'
                value={title}
                multiline
                className={classes.textField}
                margin='normal'
                fullWidth
              />

              <TextField
                label='Description'
                placeholder='e.g.: Finish biology research'
                name='description'
                value={description}
                className={classes.textField}
                margin='normal'
                fullWidth
                required
              />

              <TextField
                label='Estimated Time in days'
                type='number'
                name='estTime'
                placeholder='e.g.: 3'
                value={estTime}
                className={classes.textField}
                InputProps={{ inputProps: { min: 1 } }}
                margin='normal'
                fullWidth
                required
              />

              {modalType === 'put' && (
                <Fragment>
                  <RadioGroup
                    onChange={e => this.handleStatusChange(e)}
                    aria-label='status'
                    name='status'
                    value={done}
                    className={classes.radioButton}
                  >
                    <FormControlLabel
                      control={<Radio color='primary' />}
                      label='Done'
                      name='true'
                      value='true'
                    />
                    <FormControlLabel
                      control={<Radio color='secondary' />}
                      label='Pending'
                      name='false'
                      value='false'
                    />
                  </RadioGroup>
                </Fragment>
              )}

              <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <DateTimePicker
                  label='Due date:'
                  ampm={false}
                  value={dueDate}
                  onChange={this.handleDateChange}
                  className={classes.dateField}
                  animateYearScrolling={false}
                  required
                  disablePast
                  format={'dd/MM/yyyy HH:mm'}
                  showTodayButton
                  fullWidth
                />
              </MuiPickersUtilsProvider>

              <TagsContainer>
                <TagsInput
                  value={this.state.tags}
                  onlyUnique
                  maxTags={6}
                  inputProps={{
                    placeholder: 'Add a tag'
                  }}
                  onChange={this.handleTagsChange}
                />
              </TagsContainer>
            </form>
          </DialogContent>
          <DialogActions className={classes.actions}>
            <Button type='submit' form='todoinfo' color='primary'>
              {modalType === 'put' ? 'Update' : 'Insert'}
            </Button>
            <Button onClick={() => closeModal()} color='secondary'>
              Close
            </Button>
          </DialogActions>
        </Dialog>
      </Fragment>
    )
  }
}

const mapStateToProps = state => ({
  todos: state.todo.todos,
  isModalOpen: state.todo.isModalOpen,
  modalType: state.todo.modalType,
  currentTodo: state.todo.currentTodo
})

const mapDispatchToProps = dispatch => bindActionCreators(TodoActions, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withMobileDialog()(withStyles(useStyles)(Modal)))

Modal.propTypes = {
  todos: PropTypes.array.isRequired,
  isModalOpen: PropTypes.bool.isRequired,
  modalType: PropTypes.string.isRequired,
  currentTodo: PropTypes.object.isRequired
}

export { Modal }
