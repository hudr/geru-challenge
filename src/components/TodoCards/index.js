import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'

import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import format from 'date-fns/format'

import * as TodoActions from '../../redux/actions/todo'

import {
  StyledContainer,
  CardContainer,
  ContainerTitle,
  CardTitle,
  ContainerActionButtons,
  ActionButton,
  CardDescription,
  InformationContainer,
  CardStatus,
  TagContainer,
  CardTag,
  StyledFilterContainer,
  StyledForm
} from './styled'

class TodoCards extends Component {
  constructor(props) {
    super(props)
    this.state = {
      done: 'All',
      myTodos: []
    }
  }

  handleStatusChange(e) {
    this.setState({
      done: e.target.value
    })
  }

  render() {
    const { todos, openModal, removeTodo } = this.props
    const { done } = this.state

    const todoAux =
      done === 'All' ? [...todos] : todos.filter(todo => todo.done === done)

    return (
      <Fragment>
        <StyledFilterContainer>
          <StyledForm>
            <select onChange={this.handleStatusChange.bind(this)} value={done}>
              <option value='All'>All</option>
              <option value='false'>Pending</option>
              <option value='true'>Done</option>
            </select>
          </StyledForm>
        </StyledFilterContainer>
        <StyledContainer>
          {todoAux.map(todo => (
            <CardContainer key={todo.id}>
              <ContainerTitle>
                <CardTitle>{todo.title}</CardTitle>
                <ContainerActionButtons>
                  <ActionButton onClick={() => openModal('put', todo)}>
                    Edit
                  </ActionButton>
                  <ActionButton
                    bgColor='#cc0000'
                    onClick={() => removeTodo(todo.id)}
                  >
                    Remove
                  </ActionButton>
                </ContainerActionButtons>
              </ContainerTitle>
              <CardDescription>{todo.description}</CardDescription>
              <InformationContainer>
                <div>
                  <div>
                    <strong>Status: </strong>
                    {todo.done === 'true' ? (
                      <CardStatus fontColor={'green'}>Done!</CardStatus>
                    ) : (
                      <CardStatus>Pending.</CardStatus>
                    )}
                  </div>
                  <div>
                    <strong>Created:</strong>{' '}
                    {format(new Date(todo.createDate), 'dd/MM/yy HH:mm')}
                  </div>
                </div>
                <div>
                  <div>
                    <strong>Est.:</strong> {todo.estTime} day(s).
                  </div>
                  <div>
                    <strong>Due:</strong>{' '}
                    {format(new Date(todo.dueDate), 'dd/MM/yy HH:mm')}
                  </div>
                </div>
              </InformationContainer>
              <TagContainer>
                {todo.tags.map((tag, i) => (
                  <CardTag key={i}>#{tag.toLowerCase()}</CardTag>
                ))}
              </TagContainer>
            </CardContainer>
          ))}
        </StyledContainer>
      </Fragment>
    )
  }
}

const mapStateToProps = state => ({
  todos: state.todo.todos,
  modalType: state.todo.modalType
})

const mapDispatchToProps = dispatch => bindActionCreators(TodoActions, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoCards)

TodoCards.propTypes = {
  todos: PropTypes.array.isRequired,
  modalType: PropTypes.string
}

export { TodoCards }
