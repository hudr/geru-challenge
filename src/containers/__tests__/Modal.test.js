import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Modal } from '../../components/Modal'

const props = {
  todos: [],
  isModalOpen: false,
  modalType: '',
  currentTodo: {},
  classes: {}
}

describe('<Modal />', () => {
  it('renders without crashing given props', () => {
    const wrapper = shallow(<Modal {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
