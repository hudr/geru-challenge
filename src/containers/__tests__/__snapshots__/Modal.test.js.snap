// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> renders without crashing given props 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Dialog))
    aria-labelledby="responsive-dialog-title"
    open={false}
  >
    <WithStyles(ForwardRef(DialogTitle))
      id="responsive-dialog-title"
    >
      Fill your todo info :)
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))>
        Don't forget to add tags to your todos, they're important to filter your data later.
      </WithStyles(ForwardRef(DialogContentText))>
      <form
        autoComplete="off"
        id="todoinfo"
        onChange={[Function]}
        onSubmit={[Function]}
      >
        <WithStyles(ForwardRef(TextField))
          fullWidth={true}
          label="Title"
          margin="normal"
          multiline={true}
          name="title"
          placeholder="e.g.: Homework"
          required={true}
          value=""
        />
        <WithStyles(ForwardRef(TextField))
          fullWidth={true}
          label="Description"
          margin="normal"
          name="description"
          placeholder="e.g.: Finish biology research"
          required={true}
          value=""
        />
        <WithStyles(ForwardRef(TextField))
          fullWidth={true}
          label="Estimated Time in days"
          margin="normal"
          name="estTime"
          placeholder="e.g.: 3"
          required={true}
          type="number"
          value=""
        />
        <MuiPickersUtilsProvider
          utils={[Function]}
        >
          <PickerWithState
            allowKeyboardControl={true}
            ampm={false}
            animateYearScrolling={false}
            disablePast={true}
            format="dd/MM/yyyy HH:mm"
            fullWidth={true}
            invalidDateMessage="Invalid Date Format"
            label="Due date:"
            maxDate={2100-01-01T00:00:00.000Z}
            maxDateMessage="Date should not be after maximal date"
            minDate={1900-01-01T00:00:00.000Z}
            minDateMessage="Date should not be before minimal date"
            onChange={[Function]}
            openTo="date"
            orientation="portrait"
            required={true}
            showTabs={true}
            showTodayButton={true}
            value={2019-09-14T20:27:55.762Z}
            views={
              Array [
                "year",
                "date",
                "hours",
                "minutes",
              ]
            }
            wider={true}
          />
        </MuiPickersUtilsProvider>
        <styled.div>
          <TagsInput
            addKeys={
              Array [
                9,
                13,
              ]
            }
            addOnBlur={false}
            addOnPaste={false}
            className="react-tagsinput"
            disabled={false}
            focusedClassName="react-tagsinput--focused"
            inputProps={Object {}}
            maxTags={-1}
            onChange={[Function]}
            onlyUnique={false}
            pasteSplit={[Function]}
            preventSubmit={true}
            removeKeys={
              Array [
                8,
              ]
            }
            renderInput={[Function]}
            renderLayout={[Function]}
            renderTag={[Function]}
            tagDisplayProp={null}
            tagProps={
              Object {
                "className": "react-tagsinput-tag",
                "classNameRemove": "react-tagsinput-remove",
              }
            }
            validationRegex={/\\.\\*/}
            value={Array []}
          />
        </styled.div>
      </form>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        form="todoinfo"
        type="submit"
      >
        Insert
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="secondary"
        onClick={[Function]}
      >
        Close
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
</Fragment>
`;
